Compilation fails with Intel OneApi MKL 2024.1

**What did you expect to happen?**
I expected the build script to find the includes and libraries to link, including the linking commands.

**What has happened instead?**

```
/usr/bin/ld: /home/lars/Git/AMICI/build/libamici.a(cblas.cpp.o): in function `amici::amici_dgemm(amici::BLASLayout, amici::BLASTranspose, amici::BLASTranspose, int, int, int, double, double const*, int, double const*, int, double, double*, int)':
/home/lars/Git/AMICI/src/cblas.cpp:27:(.text+0x5): undefined reference to `cblas_dgemm'
...
```

**To Reproduce**
*Steps to reproduce the behavior*

```
sudo apt install intel-oneapi-mkl-devel
./scripts/buildSuiteSparse.sh
./scripts/buildSundials.sh
./scripts/buildAmici.sh
```


**AMICI version and system environment**
 - OS and version: Ubuntu 24.04
 - AMICI interface: C++, but actually I want to build the amici Python Interface via `pip install amici`
 - AMICI version: master / v0.25.0 (tested on both)
 - Additional information: gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0, python 3.12, intel oneapi mkl (blas)
 
```
MKLROOT=/opt/intel/oneapi/mkl/2024.1
LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.12/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/mkl/2024.1/lib/:/opt/intel/oneapi/compiler/2024.1/lib
LD_LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.12/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/mkl/2024.1/lib:/opt/intel/oneapi/compiler/2024.1/opt/compiler/lib:/opt/intel/oneapi/compiler/2024.1/lib:/usr/local/cuda-12.4/lib64
CPATH=/opt/intel/oneapi/tbb/2021.12/env/../include:/opt/intel/oneapi/mkl/2024.1/include:/opt/intel/oneapi/mkl/2024.1/include:/opt/intel/oneapi/tbb/2021.12/env/../include:/opt/intel/oneapi/mkl/2024.1/include:

 ```

**How to fix**
*Do you know how to resolve the problem?*

```
find_package(MKL CONFIG REQUIRED PATHS $ENV{MKLROOT})
  target_link_libraries(BLAS INTERFACE
    mkl_scalapack_ilp64
    mkl_cdft_core
    mkl_intel_ilp64
    mkl_intel_thread
    mkl_core
    mkl_blacs_intelmpi_ilp64
    iomp5
  ) 
```

*Can you submit a pull request?*
I don't know enough about the amici library and the reasons for the design of the `AmiciFindBLAS.cmake` file to open a pull request. Maybe after some feedback, I could open one. I would try to rely more on
`find(BLAS ...)`[https://cmake.org/cmake/help/latest/module/FindBLAS.html] or `find(MKL ...)`[https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-linux/2023-2/cmake-config-for-onemkl.html]

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -135,7 +135,7 @@ endif()
 set(VENDORED_SUNDIALS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/ThirdParty/sundials)
 set(SUNDIALS_PRIVATE_INCLUDE_DIRS "${VENDORED_SUNDIALS_DIR}/src")
 # Handle different sundials build/install dirs, depending on whether we are
-#  building the Python extension only or the full C++ interface
+# building the Python extension only or the full C++ interface
 if(AMICI_PYTHON_BUILD_EXT_ONLY)
   set(VENDORED_SUNDIALS_BUILD_DIR ${CMAKE_CURRENT_SOURCE_DIR})
   set(VENDORED_SUNDIALS_INSTALL_DIR ${VENDORED_SUNDIALS_BUILD_DIR})

--- a/cmake/AmiciFindBLAS.cmake
+++ b/cmake/AmiciFindBLAS.cmake
@@ -12,18 +12,32 @@ set_property(CACHE BLAS PROPERTY STRINGS "CBLAS" "MKL" "ACCELERATE")
 
 if(${BLAS} STREQUAL "MKL" OR DEFINED ENV{MKLROOT})
   if(DEFINED ENV{MKLROOT})
-    # This is set by Environment Modules
-    message(STATUS "Using MKL_INCDIR and MKL_LIB from environment module")
     set(BLAS
         "MKL"
         CACHE STRING "BLAS library to use" FORCE)
-    set(BLAS_INCLUDE_DIRS
-        "$ENV{MKL_INCDIR}"
-        CACHE STRING "" FORCE)
-    set(BLAS_LIBRARIES
-        "$ENV{MKL_LIB}"
-        CACHE STRING "" FORCE)
+
+    # Was MKLROOT set by /opt/intel/oneapi/setvars.sh? then cmake will find it
+    message(STATUS "Trying to find BLAS based on MKLROOT=$ENV{MKLROOT}")
+    # give the user the option to override the BLA_VENDOR
+    if(NOT DEFINED BLA_VENDOR)
+      set(BLA_VENDOR Intel10_64lp)
+    endif()
+    find_package(BLAS)
+    if(BLAS_FOUND)
+      message(STATUS "Found BLAS via FindBLAS and MKLROOT")
+    else()
+      # This is set by Environment Modules and might not be compatible with
+      # FindBLAS
+      message(STATUS "Using MKL_INCDIR and MKL_LIB from environment module")
+      set(BLAS_INCLUDE_DIRS
+          "$ENV{MKL_INCDIR}"
+          CACHE STRING "" FORCE)
+      set(BLAS_LIBRARIES
+          "$ENV{MKL_LIB}"
+          CACHE STRING "" FORCE)
+    endif()
   else()
+    message(STATUS "BLAS is set to MKL, but MKLROOT is not set.")
     set(BLAS_INCLUDE_DIRS
         ""
         CACHE STRING "")
@@ -79,17 +93,17 @@ endif()
 # Create an imported target
 if(NOT TARGET BLAS::BLAS)
   add_library(BLAS INTERFACE)
-  set_target_properties(BLAS PROPERTIES
-    INTERFACE_INCLUDE_DIRECTORIES "${BLAS_INCLUDE_DIRS}"
-    INTERFACE_LINK_LIBRARIES "${BLAS_LIBRARIES}")
+  set_target_properties(
+    BLAS PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${BLAS_INCLUDE_DIRS}"
+                    INTERFACE_LINK_LIBRARIES "${BLAS_LIBRARIES}")
   add_library(BLAS::BLAS ALIAS BLAS)
   if("${PROJECT_NAME}" STREQUAL "amici")
-      install(TARGETS BLAS EXPORT BLAS)
-      export(EXPORT BLAS NAMESPACE BLAS::)
-      install(
-        EXPORT BLAS
-        DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/Amici"
-        NAMESPACE BLAS::)
+    install(TARGETS BLAS EXPORT BLAS)
+    export(EXPORT BLAS NAMESPACE BLAS::)
+    install(
+      EXPORT BLAS
+      DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/Amici"
+      NAMESPACE BLAS::)
   endif()
 
   # legacy python package environment variables:

--- a/cmake/AmiciFindBLAS.cmake
+++ b/cmake/AmiciFindBLAS.cmake
@@ -22,6 +22,7 @@ if(${BLAS} STREQUAL "MKL" OR DEFINED ENV{MKLROOT})
     if(NOT DEFINED BLA_VENDOR)
       set(BLA_VENDOR Intel10_64lp)
     endif()
+    message(STATUS "Trying FindBLAS with BLA_VENDOR=${BLA_VENDOR}")
     find_package(BLAS)
     if(BLAS_FOUND)
       message(STATUS "Found BLAS via FindBLAS and MKLROOT")
@@ -56,6 +57,7 @@ elseif(
 
   if(APPLE AND (NOT DEFINED BLA_VENDOR OR BLA_VENDOR STREQUAL "All"))
     set(BLA_VENDOR Apple)
+    message(STATUS "Trying FindBLAS with BLA_VENDOR=${BLA_VENDOR}")
     find_package(BLAS)
     if(BLAS_FOUND)
       message(STATUS "Found Apple Accelerate BLAS")
@@ -73,6 +75,7 @@ elseif(
 
   endif()
   if(NOT BLAS_FOUND)
+    message(STATUS "Trying FindBLAS with BLA_VENDOR=${BLA_VENDOR}")
     find_package(BLAS)
     if(BLAS_FOUND)
       message(STATUS "Found BLAS via FindBLAS")

--- a/cmake/cmakelang-tools.cmake
+++ b/cmake/cmakelang-tools.cmake
@@ -9,7 +9,9 @@ set(ALL_CMAKE_FILES
     tests/cpp/unittests/CMakeLists.txt
     ${CMAKE_MODULE_PATH}/cmakelang-tools.cmake
     ${CMAKE_MODULE_PATH}/clang-tools.cmake
+    ${CMAKE_MODULE_PATH}/AmiciFindBLAS.cmake
     ${CMAKE_MODULE_PATH}/version.cmake)
+
 list(JOIN ALL_CMAKE_FILES " " ALL_CMAKE_FILES)
 
 # --- cmake-format ---

--- a/swig/CMakeLists.txt
+++ b/swig/CMakeLists.txt
@@ -20,7 +20,10 @@ find_package(
   Python3
   COMPONENTS Interpreter Development NumPy
   REQUIRED)
-message(STATUS "Found numpy ${Python3_NumPy_VERSION} include dir ${Python3_NumPy_INCLUDE_DIRS}")
+message(
+  STATUS
+    "Found numpy ${Python3_NumPy_VERSION} include dir ${Python3_NumPy_INCLUDE_DIRS}"
+)
 set(AMICI_INTERFACE_LIST
     ${CMAKE_CURRENT_SOURCE_DIR}/amici.i
     ${CMAKE_CURRENT_SOURCE_DIR}/edata.i

